fairscale/nn/data_parallel/sharded_ddp.py

```diff
@@ -604,3 +598,46 @@ def _try_consume_work_handle(self) -> None:
             work_handle = self._work_handles.popleft()
             if work_handle.callback is not None:
                 work_handle.callback()
+
+    def _attach_failsafe_hook_to_inputs(self, *inputs: Any) -> Any:
```

Comment:
this is a failsafe solution for when people are using reduce buffers, but not all parameters are being used. It should typically work in vision (inputs are floats) but not in NLP (inputs are tokens), in that case if parameters require grads but are not used ShardedDDP will assert as before

Reply:
removed, it turns out that one cannot assume that this will be called last